File: internal/repository/link.go
package repository

import (
	"context"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

type Link struct {
	ID        int       `json:"id"`
	Slug      string    `json:"slug"`
	TargetURL string    `json:"target_url"`
	Clicks    int       `json:"clicks"`
	CreatedAt time.Time `json:"created_at"`
}

type LinkRepository struct {
	pool *pgxpool.Pool
}

func New(pool *pgxpool.Pool) *LinkRepository {
	return &LinkRepository{pool: pool}
}

func (r *LinkRepository) GetBySlug(ctx context.Context, slug string) (*Link, error) {
	link := &Link{}
	err := r.pool.QueryRow(ctx,
		`SELECT id, slug, target_url, clicks, created_at FROM links WHERE slug = $1`,
		slug,
	).Scan(&link.ID, &link.Slug, &link.TargetURL, &link.Clicks, &link.CreatedAt)
	if err != nil {
		return nil, err
	}
	return link, nil
}

func (r *LinkRepository) IncrementClicks(ctx context.Context, slug string) error {
	_, err := r.pool.Exec(ctx,
		`UPDATE links SET clicks = clicks + 1 WHERE slug = $1`,
		slug,
	)
	return err
}

func (r *LinkRepository) Create(ctx context.Context, slug, targetURL string) (*Link, error) {
	link := &Link{}
	err := r.pool.QueryRow(ctx,
		`INSERT INTO links (slug, target_url) VALUES ($1, $2)
		 RETURNING id, slug, target_url, clicks, created_at`,
		slug, targetURL,
	).Scan(&link.ID, &link.Slug, &link.TargetURL, &link.Clicks, &link.CreatedAt)
	if err != nil {
		return nil, err
	}
	return link, nil
}

func (r *LinkRepository) List(ctx context.Context) ([]Link, error) {
	rows, err := r.pool.Query(ctx,
		`SELECT id, slug, target_url, clicks, created_at FROM links ORDER BY created_at DESC`,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var links []Link
	for rows.Next() {
		var l Link
		if err := rows.Scan(&l.ID, &l.Slug, &l.TargetURL, &l.Clicks, &l.CreatedAt); err != nil {
			return nil, err
		}
		links = append(links, l)
	}
	return links, nil
}

func (r *LinkRepository) Delete(ctx context.Context, slug string) error {
	_, err := r.pool.Exec(ctx,
		`DELETE FROM links WHERE slug = $1`,
		slug,
	)
	return err
}

File: src/repository/link.ts
import sql from "../db";

export interface Link {
  id: number;
  slug: string;
  target_url: string;
  clicks: number;
  created_at: Date;
}

export interface LinkStats {
  total_links: number;
  total_clicks: number;
  active_links: number;
  top_links: Link[];
  recent_links: Link[];
}

export async function getBySlug(slug: string): Promise<Link | null> {
  const [link] = await sql<Link[]>`
    SELECT * FROM links WHERE slug = ${slug}
  `;
  return link ?? null;
}

export async function incrementClicks(slug: string): Promise<void> {
  await sql`UPDATE links SET clicks = clicks + 1 WHERE slug = ${slug}`;
}

export async function create(slug: string, target_url: string): Promise<Link> {
  const [link] = await sql<Link[]>`
    INSERT INTO links (slug, target_url)
    VALUES (${slug}, ${target_url})
    RETURNING *
  `;
  return link;
}

export async function list(): Promise<Link[]> {
  return sql<Link[]>`SELECT * FROM links ORDER BY created_at DESC`;
}

export async function stats(): Promise<LinkStats> {
  const [summary] = await sql<{
    total_links: number;
    total_clicks: number;
    active_links: number;
  }[]>`
    SELECT
      COUNT(*)::int AS total_links,
      COALESCE(SUM(clicks), 0)::int AS total_clicks,
      COUNT(*) FILTER (WHERE clicks > 0)::int AS active_links
    FROM links
  `;

  const topLinks = await sql<Link[]>`
    SELECT *
    FROM links
    ORDER BY clicks DESC, created_at DESC
    LIMIT 5
  `;

  const recentLinks = await sql<Link[]>`
    SELECT *
    FROM links
    ORDER BY created_at DESC
    LIMIT 5
  `;

  return {
    total_links: summary?.total_links ?? 0,
    total_clicks: summary?.total_clicks ?? 0,
    active_links: summary?.active_links ?? 0,
    top_links: topLinks,
    recent_links: recentLinks,
  };
}

export async function remove(slug: string): Promise<void> {
  await sql`DELETE FROM links WHERE slug = ${slug}`;
}
